components/use-paid-mutation.js

```diff
@@ -23,39 +25,144 @@ export function usePaidMutation (mutation,
   const [getPaidAction] = useLazyQuery(GET_PAID_ACTION, {
     fetchPolicy: 'network-only'
   })
+  const [retryPaidAction] = useMutation(RETRY_PAID_ACTION)
   const waitForWalletPayment = useWalletPayment()
   const invoiceHelper = useInvoice()
   const waitForQrPayment = useQrPayment()
   const client = useApolloClient()
   // innerResult is used to store/control the result of the mutation when innerMutate runs
   const [innerResult, setInnerResult] = useState(result)
+  const { me } = useMe()
 
-  const waitForPayment = useCallback(async (invoice, { alwaysShowQROnFailure = false, persistOnNavigate = false, waitFor }) => {
-    let walletError
-    const start = Date.now()
+  const senderWallets = useEnabledWallets().map(w => {
+    return { ...w, failed: false }
+  })
+
+  const addPayError = (e, rest) => ({
+    ...rest,
+    payError: e,
+    error: e instanceof InvoiceCanceledError && e.actionError ? e : undefined
+  })
+
+  const waitForActionPayment = useCallback(async (invoice, { alwaysShowQROnFailure = false, persistOnNavigate = false, waitFor }, originalResponse, action) => {
+    const walletErrors = []
+    let response = originalResponse
+    let invoiceUsed = false
+
+    const cancelInvoice = async () => {
+      try {
+        invoiceUsed = true
+        await invoiceHelper.cancel(invoice)
+        console.log('old invoice canceled')
+      } catch (err) {
+        console.error('could not cancel old invoice', err)
```

Comment:
i am not sure, won't this throw if the invoice was cancelled twice or is expired too?

Reply:
Canceling the previous invoice is pretty important since we don't want them to accidentally pay us twice. This code doesn't run exclusively for p2p zaps where we wouldn't be able to settle incoming invoices ourselves without the receiver settling first; **this runs for every paid action**. Our node would absolutely settle an old invoice even though they already paid for the action.

Reply:
So, this is an issue for optimistic actions, that can fail after the invoice is settled, but not for pessimistic actions that use hodl invoice and set the FAILED state on settling, is this correct?